go/vt/vtctl/workflow/server.go

```diff
@@ -850,6 +852,207 @@ func (s *Server) Materialize(ctx context.Context, ms *vtctldatapb.MaterializeSet
 	return mz.startStreams(ctx)
 }
 
+// MaterializeAddTables adds specified tables to the existing workflow.
+func (s *Server) MaterializeAddTables(ctx context.Context, req *vtctldatapb.MaterializeAddTablesRequest) error {
+	if len(req.TableSettings) == 0 {
+		return vterrors.Errorf(vtrpcpb.Code_FAILED_PRECONDITION, "no tables found in the request")
+	}
+
+	targets, err := s.ts.FindAllShardsInKeyspace(ctx, req.Keyspace, nil)
+	if err != nil {
+		return err
+	}
+	targetShardInfos := maps.Values(targets)
+
+	// This will be helpful to find duplicate tables.
+	tableSet := sets.New[string]()
+	for _, ts := range req.TableSettings {
+		tableSet.Insert(ts.TargetTable)
+	}
+
+	// Store the ReadVReplicationWorkflow response for later use.
+	readVReplicationWorkflowResp := make(map[string]*tabletmanagerdatapb.ReadVReplicationWorkflowResponse)
+	var (
+		mu             sync.Mutex
+		sourceKeyspace string
+		workflowType   binlogdatapb.VReplicationWorkflowType
+	)
+
```

Comment:
We need to get a named workflow lock and a target keyspace lock here. Otherwise there's no concurrency control in place and behavior is undefined.

Reply:
done, thanks!